### PR TITLE
fix(eslint-config): set parser, don't fail the build on minimal eslint-loader errors

### PR DESCRIPTION
**Pull request by pieh**
## Description

https://github.com/gatsbyjs/gatsby/pull/29109 was not sufficent. This also specify `parser` (by default it uses https://github.com/eslint/espree - but it seems like version we/eslint (also version of eslint we use) doesn't support optional chaining `?.` - used version is `espree@^6`, and according to https://github.com/eslint/espree/blob/master/CHANGELOG.md support was added in `espree@^7`).

Other part is `emitWarning: true`, which _should_ prevent errors coming from `eslint-loader` from blocking users - per https://github.com/webpack-contrib/eslint-loader#noemitonerrorsplugin

> `NoEmitOnErrorsPlugin` is now automatically enabled in webpack 4, when mode is either unset, or set to production. So even ESLint warnings will fail the build. No matter what error settings are used for `eslint-loader`, except if `emitWarning` enabled. 

Current:
Screenshot 2021-01-21 at 20 30 47

With `emitWarning: true` - so it still is confusing, and we need to fix those, but at least it doesn't prevent webpack from compiling (or rather emitting I guess):
Screenshot 2021-01-21 at 20 30 36

It prooved to be whack'a'mole with this minimal eslint config for fast-refresh rules, so there are few things to consider here:
 - maybe we should disable it (for now)?
 - lack of tests for this code path doesn't inspire confidence

## Related Issues

https://github.com/gatsbyjs/gatsby/issues/29105#issuecomment-764856608

Fixes https://github.com/gatsbyjs/gatsby/issues/29133

### CODE REVIEW

**Comment by theskillwithin** [comment]
I am experiencing this as well as long with `Parsing error: Unexpected token =`  on old class components

**Comment by mikaelkristiansson** [comment]
what is the progress of this?

File: packages/gatsby/src/utils/eslint-config.ts [hunk 1]
```diff
@@ -5,10 +5,22 @@ import path from "path"
...
+export const eslintRequiredConfig: EslintLoaderOptions = {
   rulePaths: [eslintRulePaths],
   useEslintrc: false,
+  // this forces any errors to be webpack warnings, minimal eslint-loader
+  // should NEVER prevent bundle from building. Our eslint config for it
+  // only contain rules set to warnings, but loader itself can emit errors
+  // if parsing failed etc.
+  emitWarning: true,
   baseConfig: {
+    parser: require.resolve(`babel-eslint`),
```
Comment (wardpeet) [inline packages/gatsby/src/utils/eslint-config.ts:23]: Shouldn't we add it to the package.json?
```suggestion
    parser: require.resolve(`babel-eslint`),
```

File: packages/gatsby/src/utils/eslint-config.ts [hunk 2]
```diff
@@ -5,10 +5,22 @@ import path from "path"
...
     parserOptions: {
```
Comment (theskillwithin) [inline packages/gatsby/src/utils/eslint-config.ts:24]: for ecmaVersion I set to 2020.  any reason why its 2018 here?
```
  parserOptions: {
    ecmaVersion: 2020,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true,
    },
  },
  ```

File: packages/gatsby/src/utils/eslint-config.ts (hunk 1 above)
Comment (pieh) [reply packages/gatsby/src/utils/eslint-config.ts:23]: we already have it - https://github.com/gatsbyjs/gatsby/blob/7b6454b753f1098ee76b21a18cc1f6d16a1fe112/packages/gatsby/package.json#L34

**Comment by vladar** [comment]
I don't think it fixes issue #29133 

The problem is that we add our required config here:

https://github.com/gatsbyjs/gatsby/blob/7b6454b753f1098ee76b21a18cc1f6d16a1fe112/packages/gatsby/src/utils/webpack.config.js#L737-L743

But after #29109 this required config has `useEslintrc: false`. So we effectively disable `eslintrc` with our new settings, i.e. even if the file exists, the loader ignores it with our config.

So to recap - this PR indeed fixes the problem with parser but not the `eslintrc` part.

As for me, it sounds like the only good way to fix this is to have two eslint loaders - one for normal eslint setup - and another one - for specific fast-refresh rules. That separate loader can have any config, e.g. we can safely disable eslintrc and inline comments for it.

**Comment by pieh** [comment]
Closing this in favour of https://github.com/gatsbyjs/gatsby/pull/29317 which is more robust
